torch/tensor.py

```diff
@@ -266,30 +266,31 @@ def repeat(self, *sizes):
         # If args == (torch.Size,), then we need to unpack the tuple
         if len(sizes) == 1 and isinstance(sizes[0], torch.Size):
             sizes = sizes[0]
+
         repeats = list(sizes)
-        result = self.new()
-        src = self.contiguous()
 
-        if len(repeats) < src.dim():
+        if len(repeats) < self.dim():
             raise ValueError('Number of dimensions of repeat dims can not be '
                              'smaller than number of dimensions of tensor')
 
-        xtensor = src.new().set_(src)
-        xsize = list(xtensor.size())
-        for i in _range(len(repeats) - src.dim()):
-            xsize = [1] + xsize
+        # Add new leading dimensions to the tensor if the
+        # number of target dimensions is larger than the
+        # number of source dimensions.
+        num_new_dimensions = len(repeats) - self.dim()
+        padded_size = [1] * num_new_dimensions + list(self.size())
+        target_size = torch.Size([a * b for a, b in zip(padded_size, repeats)])
+
+        xtensor = self.new().set_(self)
+        xtensor = xtensor.expand(padded_size)
 
-        size = torch.Size([a * b for a, b in zip(xsize, repeats)])
-        xtensor.resize_(torch.Size(xsize))
-        result.resize_(size)
+        result = self.new()
+        result.resize_(target_size)
         urtensor = result.new(result)
         for i in _range(xtensor.dim()):
             urtensor = urtensor.unfold(i, xtensor.size(i), xtensor.size(i))
-        for i in _range(urtensor.dim() - xtensor.dim()):
-            xsize = [1] + xsize
-        xtensor.resize_(torch.Size(xsize))
-        xxtensor = xtensor.expand_as(urtensor)
-        urtensor.copy_(xxtensor)
+
+        urtensor.copy_(xtensor.expand_as(urtensor))
```

Comment:
Is this `expand_as` really needed? I'm not 100% how these unfolds let us achieve repeating, but it seems that there's no guarantee that `xtensor` is contiguous, so if this wasn't a no-op I would expect it to break.

Reply:
To be perfectly frank I have no idea what is going on there. I think `urtensor` gets some new dimensions beyond `xtensor`, then the copy semantics somehow make it work after expanding.